### PR TITLE
Add benchmarks

### PR DESCRIPTION
**Pull request by elichai**
I wanted to get some feedback before continuing this.

I started implementing benchmarks so that changes for "performance" can be actually evaluated (for example, I want to try and replace a bunch of for loops with iterators and see if it has any effect).

### CODE REVIEW

**Comment by apoelstra** [comment]
Nice. I'm in favor of this.

A bit weird that we had the `unstable` and `test` features, but AFAICT these weren't actually used? Maybe there used to be benchmarks in this library a very long time ago.

File: src/blockdata/block.rs [hunk 1]
```diff
@@ -282,3 +282,50 @@ mod tests {
...
...56339913718ce18dbb1a27bdb00220482911c851d15adcd37097dff99a9ff1f97d953bcebc528835118f447412553e014c695221028d9889862b29430278c084b5c4090b7b807b31e047bcd212ebc2c4e43fc0e3c52103160949a7c8c81f2c25d7763f57eb1cb407d867c5b7c290331bd2dc4b1182c6d32103fbef3b60914bda9173765902013a251ec89450c75d0b5a96a143db1dabf98d9553aeffffffff0220e8891c0100000017a914d996715e081c50f8f6b1b4e7fb6ca214f9924fdf87809698000000000017a9145611d812263f32960228cb5f85329bce4770a218870000000001000000017720507dcbe6c69f652b0c0ce19406f482372d1a8abc05d45fb7acf97fb80eec00000000fdfe00004830450221009821d8e117de44b1202c829c0f5063997acf007cf9b561c6fb8d1212cddb6c40022010ff5067b0d9d4eca2da0ceb876e9a16f1a2142da866d3042a7bae8968813e8001483045022100dea759d14a8a1c5da5f3dcc5509871aaa2c1e3be03752c1b858d80fa4227163702205183d70cc28dcb6df9b037714c8b6442ef84e0ddce07711a30c731e9f0925090014c695221028d70ea66fe7a7def282df7b2b498007e5072933e42c18f63ce85975dcbcf1a8821037e8f842b1e47e21d88002c5aab2559212a4c2c9dbe5ef5347f2a29afd0510ec1210251259cb9fd4f6206488408286e4475c9c9fe887e57a3e32ae4da222778a2aedf53aeffffffff023380cb020000000017a9143b5a7e85b22656a34d43187ac8dd09acd7109d2487809698000000000017a914b9b4b555f594a34deec3ad61d5c5f3738b17ee158700000000";
```
Comment (stevenroose) [inline src/blockdata/block.rs:293]: Perhaps add a line of comment on this block. Why was it picked? Is it mainnet? Is it large? ...

File: src/blockdata/transaction.rs [hunk 2]
```diff
@@ -1233,3 +1233,60 @@ mod tests {
...
 
+#[cfg(all(test, feature = "unstable"))]
+mod benches {
+    use super::Transaction;
+    use EmptyWrite;
+    use consensus::{deserialize, Encodable};
+    use hashes::hex::FromHex;
+    use test::{black_box, Bencher};
+
+    const SOME_TX: &'static str = "0100000001a15d57094aa7a21a28cb20b59aab8fc7d1149a3bdbcddba9c622e4f5f6a99ece010000006c493046022100f93bb0e7d8db7bd46e40132d1f8242026e045f03a0efe71bbb8e3f475e970d790221009337cd7f1f929f00cc6ff01f03729b069a7c21b59b1736ddfee5db5946c5da8c0121033b9b137ee87d5a812d6f506efdd37f0affa7ffc310711c06c7f3e097c9447c52ffffffff0100e1f505000000001976a9140389035a9225b3839e2bbf32d826a1e222031fd888ac00000000";
```
Comment (stevenroose) [inline src/blockdata/transaction.rs:1244]: Perhaps add a line of comment on this tx. Why was it picked? Is it mainnet? Is it large? ...

File: src/lib.rs [hunk 3]
```diff
@@ -99,3 +99,33 @@ pub use util::amount::SignedAmount;
 pub use util::key::PrivateKey;
 pub use util::key::PublicKey;
 pub use util::merkleblock::MerkleBlock;
+
+#[cfg(all(test, feature = "unstable"))] use tests::EmptyWrite;
```
Comment (stevenroose) [inline src/lib.rs:103]: This can be replaced with making the `mod tests` `pub(crate) mod test`, right?

**Review by stevenroose — DISMISSED** [review]
I have nothing against this.

**Comment by apoelstra** [comment]
Needs rebase

**Review by stevenroose — APPROVED** [review]
LGTM

**Review by apoelstra — APPROVED** [review]
ack a44ba2d87850f6ab1cd02209229a2c52c4ebb173 (pending CI and local tests passing)

**Pull request merged**